sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -49,23 +50,47 @@ def _find_binning_thresholds(col_data, max_bins):
         col_data = col_data[~missing_mask]
     # The data will be sorted anyway in np.unique and again in percentile, so we do it
     # here. Sorting also returns a contiguous array.
-    col_data = np.sort(col_data)
+    sort_idx = np.argsort(col_data)
+    col_data = col_data[sort_idx]
     distinct_values = np.unique(col_data).astype(X_DTYPE)
-    if len(distinct_values) <= max_bins:
-        midpoints = distinct_values[:-1] + distinct_values[1:]
-        midpoints *= 0.5
+    if sample_weight is None:
+        if len(distinct_values) <= max_bins:
+            midpoints = distinct_values[:-1] + distinct_values[1:]
+            midpoints *= 0.5
+        else:
+            percentiles = np.linspace(0, 100, num=max_bins + 1)
+            percentiles = percentiles[1:-1]
+            midpoints = np.percentile(
+                col_data, percentiles, method="averaged_inverted_cdf"
+            )
+            assert midpoints.shape[0] == max_bins - 1
+
     else:
         # We could compute approximate midpoint percentiles using the output of
         # np.unique(col_data, return_counts) instead but this is more
         # work and the performance benefit will be limited because we
         # work on a fixed-size subsample of the full data.
+        # TO DO: check if there is a better way to implement this
+        sample_weight = sample_weight[sort_idx]
         percentiles = np.linspace(0, 100, num=max_bins + 1)
         percentiles = percentiles[1:-1]
-        midpoints = np.percentile(col_data, percentiles, method="midpoint").astype(
-            X_DTYPE
+
+        midpoints = np.array(
+            [
+                _averaged_weighted_percentile(col_data, sample_weight, percentile)
+                for percentile in percentiles
+            ]
         )
         assert midpoints.shape[0] == max_bins - 1
 
+        # Remove duplicated midpoints if they exist and shift
+        # by 0.5 if the unique points are less than distinct
+        # values
+        if np.unique(midpoints).shape[0] != midpoints.shape[0]:
+            midpoints = np.unique(midpoints)
```

Comment:
We should avoid calling `np.unique(midpoints)` repeatedly. At this stage, the input array has `max_bins` values so it's not that costly, but still, let store this is a local variable.